### PR TITLE
Introduce a new deleteEach method for Table

### LINKED ISSUE #9727: Introduce a new deleteEach method for Table
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.next
* Platform and Target: Apache, MySQL, PHP 7

### What you did

The classic example

```
     $table->addBehavior('Something');
     $table->deleteAll(['thing'=>true]);
```

### What happened

Records are deleted, but no behaviors/events are processed.

### What you expected to happen

There is a history of users opening issues about `deleteAll` failing to perform some kind of expected callback operation on records. It is a short hand method for running a single SQL statement.

Is there value in adding a `deleteEach` that would iterate over the results deleting each record individually? That would allow for the expected events to be processed.

Here's an example as a quick draft. I would add more options in the final PR.

```
     public function deleteEach($conditions)
     {
            $count = 0;
            while(true) {
                 $entity = $this->find()->where($conditions)->first();
                 if(!$entity || !$this->delete($entity)) {
                      break;
                 }
                 $count++;
            }
            return $count;
     }
```

### PR DESCRIPTION
**Pull request by thinkingmedia**
closes #9727

- adds a new deleteEach method to the Table interface
- adds beforeDeleteEach  and afterDeleteEach events

Unit tests are coming. Just wanted early review.

### CODE REVIEW

File: src/Datasource/RepositoryInterface.php [hunk 1]
```diff
@@ -115,6 +115,23 @@ public function updateAll($fields, $conditions);
     public function deleteAll($conditions);
 
     /**
+     * Delete each record individually by matches the conditions. For each
```
Comment (markstory) [inline src/Datasource/RepositoryInterface.php]: s/matches/matching/

File: src/ORM/Table.php [hunk 2]
```diff
@@ -1368,6 +1367,37 @@ public function deleteAll($conditions)
     /**
      * {@inheritDoc}
      */
+    public function deleteEach($conditions, $options = [])
+    {
+        $options = $options + [
```
Comment (ADmad) [inline src/ORM/Table.php]: $options +=  [

File: src/ORM/Table.php [hunk 3]
```diff
@@ -1368,6 +1367,37 @@ public function deleteAll($conditions)
...
+
+        $event = $this->dispatchEvent('Model.beforeDeleteEach', compact('conditions', 'options'));
+        if ($event->isStopped()) {
+            return $event->result();
+        }
+        $conditions = $event->data('conditions');
+
+        $count = 0;
+        do {
+            $entity = $this->find()->where($conditions)->first();
```
Comment (ADmad) [inline src/ORM/Table.php]: Why not fetch all entities first and then delete one by one?

File: src/ORM/Table.php [hunk 4]
```diff
@@ -1368,6 +1367,37 @@ public function deleteAll($conditions)
     /**
      * {@inheritDoc}
      */
+    public function deleteEach($conditions, $options = [])
+    {
+        $options = $options + [
+                'strict' => true,
+                'limit' => 0
+            ];
```
Comment (markstory) [inline src/ORM/Table.php]: Could be `$options += [ ... ]`

File: src/ORM/Table.php (hunk 2 above)
Comment (ADmad) [inline src/ORM/Table.php]: Also the options need to be documented.

File: src/ORM/Table.php [hunk 5]
```diff
@@ -1368,6 +1367,37 @@ public function deleteAll($conditions)
...
+            if ($entity === null) {
+                break;
+            }
+            if ($this->delete($entity, $options) === false && $options['strict']) {
+                break;
+            }
+            $count++;
+        } while ($entity !== null && ($options['limit'] == 0 || $count < $options['limit']));
+
+        return $this->dispatchEvent('Model.afterDeleteEach', compact('count'))->data('count');
```
Comment (markstory) [inline src/ORM/Table.php]: Should this event get the records that were deleted?

File: src/ORM/Table.php [hunk 6]
```diff
@@ -1368,6 +1367,37 @@ public function deleteAll($conditions)
...
+        }
+        $conditions = $event->data('conditions');
+
+        $count = 0;
+        do {
+            $entity = $this->find()->where($conditions)->first();
+            if ($entity === null) {
+                break;
+            }
+            if ($this->delete($entity, $options) === false && $options['strict']) {
```
Comment (ADmad) [inline src/ORM/Table.php]: Option name doesn't sound too intuitive for what it does.

**Review by ADmad — CHANGES_REQUESTED** [review]

File: src/ORM/Table.php (hunk 5 above)
Comment (thinkingmedia) [inline src/ORM/Table.php]: @markstory that would risk an out of memory error.

File: src/ORM/Table.php (hunk 3 above)
Comment (thinkingmedia) [inline src/ORM/Table.php]: @ADmad out of memory would be my worry. I should even include options for what fields to load.

File: src/ORM/Table.php (hunk 6 above)
Comment (thinkingmedia) [inline src/ORM/Table.php]: @ADmad how about "stopOnFailure"?

File: src/Datasource/RepositoryInterface.php [hunk 7]
```diff
@@ -115,6 +115,30 @@ public function updateAll($fields, $conditions);
...
+     * ### Options
+     *
+     * Options are passed to Table::delete() with the exception of these.
+     *
+     * - 'stopOnFailure' boolean (default True) stop deleting records should one fail to be deleted.
+     * - 'limit' integer (default 0) Limit the number of records to delete, zero means no limit.
+     *
+     * @param mixed $conditions Conditions to be used, accepts anything Query::where()
+     * can take.
+     * @param array $options The options for the updates.
```
Comment (lorenzo) [inline src/Datasource/RepositoryInterface.php]: This is not an array.  It is anything that `where()` can accept

File: src/Datasource/RepositoryInterface.php [hunk 8]
```diff
@@ -115,6 +115,30 @@ public function updateAll($fields, $conditions);
...
+     * be aware of possible long execution times, and side effects of processing
+     * events on a large collection of records.
+     *
+     * ### Options
+     *
+     * Options are passed to Table::delete() with the exception of these.
+     *
+     * - 'stopOnFailure' boolean (default True) stop deleting records should one fail to be deleted.
+     * - 'limit' integer (default 0) Limit the number of records to delete, zero means no limit.
+     *
```
Comment (lorenzo) [inline src/Datasource/RepositoryInterface.php]: The doc is missing some usage examples

File: src/ORM/Table.php (hunk 3 above)
Comment (lorenzo) [inline src/ORM/Table.php]: I would load in batches, issuing one query per record will make this function even a worse idea

**Review by lorenzo — CHANGES_REQUESTED** [review]

File: src/ORM/Table.php [hunk 9]
```diff
@@ -755,7 +754,7 @@ public function belongsTo($associated, array $options = [])
         $options += ['sourceTable' => $this];
         $association = new BelongsTo($associated, $options);
 
-        return $this->_associations->add($association->name(), $association);
+        return $this->_associations->add($association->getName(), $association);
```
Comment (ADmad) [inline src/ORM/Table.php]: Please avoid making unrelated changes in a PR. It adds noise and shifts focus from task at hand.

**Comment by thinkingmedia** [comment]
@ADmad noise has been removed from the PR
@lorenzo fixed your issues as well.
@ionas deleteEach now supports transactions

This update improves deleteEach by batch processing records. It reads the records inside a transaction (if atomic) and deletes them one by one. Uses the applyOptions feature of a query so that the user can control how records are read, including the use of a custom finder.

Thanks for all the feedback, I think it pushed me to make this better than my original attempt.

File: src/Datasource/RepositoryInterface.php [hunk 10]
```diff
@@ -115,26 +115,46 @@ public function updateAll($fields, $conditions);
...
+     * and page.
      *
      * ### Options
      *
-     * Options are passed to Table::delete() with the exception of these.
+     * - 'finder' string Defaults to 'all'. The customer finder to use when reading batches.
+     * - `atomic` boolean Defaults to true. When true the deletion happens within a transaction.
+     * - 'stopOnFailure' boolean Defaults to true. Stop deleting records when a record fails to be deleted.
+     * - 'batch' integer Defaults to 1. Number of records to read in batches, zero reads all records.
+     * - `checkRules` boolean Defaults to true. Check deletion rules before deleting the record.
```
Comment (ADmad) [inline src/Datasource/RepositoryInterface.php]: Please either use single quote or backtick consistently.

File: src/Datasource/RepositoryInterface.php [hunk 11]
```diff
@@ -115,6 +115,50 @@ public function updateAll($fields, $conditions);
...
+     * $count = $articles->deleteEach(['published' => true]);
+     * ```
+     *
+     * @param mixed $conditions Conditions to be used, accepts anything Query::where()
+     * can take.
+     * @param array|\ArrayAccess $options An array that will be passed to Query::applyOptions()
+     * @return int Count Returns the affected rows.
+     * @see \Cake\Datasource\RepositoryInterface::delete()
+     */
+    public function deleteEach($conditions, $options = []);
```
Comment (ADmad) [inline src/Datasource/RepositoryInterface.php]: Adding new method to interface in minor release will break BC.

File: src/Datasource/RepositoryInterface.php (hunk 10 above)
Comment (dereuromark) [inline src/Datasource/RepositoryInterface.php]: Personally I would still prefer to let the IDE help (autocomplete and existence check) with class constants for the option keys ;)
`RepositoryInterface::FINDER => ...`

Comment (thinkingmedia) [inline src/Datasource/RepositoryInterface.php]: @dereuromark you can't define constants in an interface. I wish I could.

File: src/Datasource/RepositoryInterface.php [hunk 12]
```diff
@@ -115,6 +115,50 @@ public function updateAll($fields, $conditions);
     public function deleteAll($conditions);
 
     /**
+     * Deletes records individually by reading batches that match the conditions. For each
```
Comment (inoas) [inline src/Datasource/RepositoryInterface.php]: Maybe now that it allows to save batches it should be called batchDelete or deleteBatch and then afterDeleteBatch and and afterDeleteBatchCommit etc... ?!

On the other hand it could also say:
`Deletes records that match the conditions individually. For each matched record Table::delete() is executed providing support for behaviors and events. You can specify a batch size to delete at once to improve performance on large amounts of records.` ... or similar

File: src/Datasource/RepositoryInterface.php [hunk 13]
```diff
@@ -115,6 +115,50 @@ public function updateAll($fields, $conditions);
...
+     *
+     * This method will *not* accept query options that alter the size of a batch, such as limit, offset
+     * and page.
+     *
+     * ### Options
+     *
+     * - 'finder' string Defaults to 'all'. The customer finder to use when reading batches.
+     * - `atomic` boolean Defaults to true. When true the deletion happens within a transaction.
+     * - 'stopOnFailure' boolean Defaults to true. Stop deleting records when a record fails to be deleted.
+     * - 'batch' integer Defaults to 1. Number of records to read in batches, zero reads all records.
```
Comment (inoas) [inline src/Datasource/RepositoryInterface.php]: `Set to 0 to read all records in one single batch.`

File: src/Datasource/RepositoryInterface.php (hunk 11 above)
Comment (ADmad) [inline src/Datasource/RepositoryInterface.php]: > So I doubt anyone is using it in a way that would be a problem.

@thinkingmedia Based on past experience we know not to make assumptions about usage regarding important classes/interfaces. As @markstory stated this change would break an official plugin itself.

BC guidelines exist for a reason.

Comment (thinkingmedia) [inline src/Datasource/RepositoryInterface.php]: @ADmad I have a fix for BC. Will update the PR soon.

File: src/Datasource/RepositoryInterface.php [hunk 14]
```diff
@@ -114,6 +118,7 @@ public function updateAll($fields, $conditions);
      */
     public function deleteAll($conditions);
 
+
```
Comment (ADmad) [inline src/Datasource/RepositoryInterface.php]: extra blank line

Comment (ADmad) [inline src/Datasource/RepositoryInterface.php]: extra blank line

File: src/ORM/Table.php [hunk 15]
```diff
@@ -1576,6 +1576,116 @@ public function deleteAll($conditions)
...
+
+        $invalid = array_intersect(array_keys($options), ['limit', 'offset', 'page']);
+        if (count($invalid) !== 0) {
+            throw new InvalidArgumentException(sprintf('Option conflicts with batch deleting: %s', implode(', ', $invalid)));
+        }
+
+        $count = 0;
+
+        $atomic = function () use ($conditions, $options, &$count) {
+
```
Comment (ADmad) [inline src/ORM/Table.php]: extra blank line

**Comment by thinkingmedia** [comment]
It seems this change has passed people's review. I'll proceed now to add unit tests.

Thanks all,

**Comment by ADmad** [comment]
@thinkingmedia Did you forget about this PR? :)

**Comment by thinkingmedia** [comment]
@ADmad I did. Sorry lol

**Comment by thinkingmedia** [comment]
I'll get back to this, this week.

**Comment by ADmad** [comment]
@thinkingmedia Seems you again forgot about this PR :smile:

**Comment by thinkingmedia** [comment]
Moved PR to: https://github.com/cakephp/cakephp/pull/10284
